### ISSUE TITLE
LD sampling

### ISSUE THREAD

**Issue by elgartmi** [privefl/bigsnpr#269]
Good Day,

I'm working on a highly admixed population cohort, so I can't use the provided LD reference panel.
My cohort has ~70k people and ~0.5M SNPs per chromosome (after quality control as recommended). 
So the snp_corr never finishes (16 cores, 128G; it dis finish for chr22 which is relatively tiny, after several days ).

I was thinking to sample ~1000 people 100 times and average the resulting correlation matrices. This works quite fast, but apparently matrix sum is not implemented for dsCMatrix/dgCMatrix (of size 500000x500000) so no way to average. 

It just throws: 
`"'forcesymmetric' long vectors not supported yet".`

Is there a way to do it with big_apply or some other way?

Thanks!

**Comment by privefl** [comment]
It seems you're approaching two problems
 - choosing an optimal set of variants
 - prediction for multiple ancestries (or at least from a multi-ethnic GWAS)
 
 I can tell by experience (I am working on similar problems) that these problems are not easy so solve. You might want to pick one at a time :')

**Comment by privefl** [comment]
I think you should not subsample and sum. This is not the same as computing the correlation matrix directly. 
And I can't see how you could save time and memory doing this.

BTW, adding two `dsCMatrix` matrices works for me, e.g.:
```r
library(Matrix)
mat1 <- rsparsematrix(100, 100, 0.1, symmetric = TRUE)
mat2 <- rsparsematrix(100, 100, 0.1, symmetric = TRUE)
mat3 <- mat1 + mat2
```

**Comment by privefl** [comment]
Any update on this?

**Comment by elgartmi** [comment]
Hi,
Decided against subsampling - couldn't bring myself to write the C code to average large matrices, and it wouldn't run otherwise even with a subsample of 1K people (for ~20M variants).

To reduce number of variants, tried top 1M variants vs clumped.

**TL;DR Clumping works as well as using top 1M variants, but is much faster; similar performance to PRSice**

So I've tried doing what you did in your paper (using top 1M) variants with correction, and also clumped using PRSice, and used only the clumped variants (I think people refer to it as LD-Clump?).
Clumping seems to work better (tried on ~9 phenotypes), although not for every case and not significantly. I am currently trying to figure out how sensitive it is to the clumping parameters.

Specifically, if you set the threshold of clumping at R=0.95 (so removing only VERY correlated SNPs), you get down to ~1.5M variants from ~20M (give or take) which the algorithm can handle, though it still takes weeks for all the steps. If you use the usual R=0.1 or some such, then its 300-500 variants and a couple of days (on a cluster, running chromosomes in parallel). The clumped runs 3-5x faster than top 1M, also is better for admixed populations.

The auto mode performs best. Grid very often does not converge.
Performance-wise, I don't see too much of a difference between LD-based and PRSice for admixed populations.

For more, tune in to our publication, to which I'll post a link here once we finish addressing the infinite number of comments we got :)

Wow, that was too long...
Michael

P.S. 
 If you implement large matrix averaging, I can also try subsampling. Statistically it's more correct than either top 1M, or clumping, I think. Although since the two seem to perform very similarly, it probably means that we are close to the true value, thus the subsampling may be not needed.

**Comment by privefl** [comment]
If the aim is prediction, then you might want to try an hybrid approach where you have HM3 + top-clumped variants (e.g. p < 1e-20).
I have also tried 3 different sets (including HM3 and clumping) in a recent preprint: https://www.biorxiv.org/content/10.1101/2021.03.29.437510v2.

For the runtime, it should take 10 min for 15K variants and 360K individuals with 15 cores.
Then the runtime is quadratic with the number of variants, but only linear with the number of individuals.
So around 30 hours for 200K variants.

**Comment by privefl** [comment]
Any update on this?
